types/underscore/index.d.ts

```diff
@@ -66,67 +67,119 @@ declare module _ {
+    // if any type in the union doesn't implement the property
+    // e.g. TypesOfUnionProperty<{ a: string } | { a: number } | { b: boolean }, 'a'> => string | number | undefined
+    type TypesOfUnionProperty<T, K extends KeysOfUnion<T>> =
+        // iterate over each type T in a type union
+        T extends unknown
+        // check whether the property is implemented by T, using Extract to get around K possibly not being in keyof T
+        ? T[Extract<keyof T, K>] extends never
+        // if T does not implement property K, add undefined to the resulting type union
+        ? undefined
+        // if T does implement property K, add type T[K] to the resulting type union
+        : T[Extract<keyof T, K>]
+        // this should never actually be evaluated since all types extend any
+        : never;
+
+    type PropertyNamesOfType<T, P> = { [K in keyof T]: T[K] extends P ? K : never }[keyof T];
+
+    type ShallowFlattenedList<T> = T extends _.List<infer TItem> ? TItem[] : T[];
+
+    // unfortunately it's not possible to recursively collapse all possible list dimensions to T[] at this time,
+    // so give up after two dimensions and require an assertion
+    // surprisingly T extends _.List<_.List<infer TItem>> isn't true when T = SomeType[][],
+    // so writing this that way doesn't work
+    type DeepFlattenedList<T> = T extends _.List<infer TItem>
+        ? TItem extends _.List<infer TInnerItem>
+        ? TInnerItem extends _.List<unknown>
+        ? unknown[]
+        : TInnerItem[]
+        : TItem[]
+        : T[];
 
     interface UnderscoreStatic {
         /**
         * Underscore OOP Wrapper, all Underscore functions that take an object
         * as the first parameter can be invoked through this function.
-        * @param key First argument to Underscore object functions.
+        * @param value First argument to Underscore object functions.
         **/
-        <T>(value: _.List<T>): Underscore<T, _.List<T>>;
-        <T>(value: Array<T>): Underscore<T, Array<T>>;
-        <T extends TypeOfDictionary<V>, V extends _.Dictionary<any> = _.Dictionary<T>>(value: V): Underscore<T, V>;
-        <T>(value: T): Underscore<T>;
+        <V>(value: V): Underscore2<V>;
 
         /* *************
         * Collections *
         ************* */
 
+        /**
+        * Iterates over a list of elements, yielding each in turn to an iterator function. The iterator is
+        * bound to the context object, if one is passed. Each invocation of iterator is called with three
+        * arguments: (element, index, list). If list is a JavaScript object, iterator's arguments will be
+        * (value, key, object). Delegates to the native forEach function if it exists.
+        * @param list Iterates over this list of elements.
+        * @param iterator Iterator function for each element `list`.
+        * @param context 'this' object in `iterator`, optional.
+        **/
+        each<T>(
```

Comment:
It would be nice to update `UnderscoreStatic` functions to use collection-centric generics instead of collection-item-centric-generics e.g. each could be reduced to a single overload that always uses the type of item it was given for both iterators and return types:

```TS
each<V extends Collection<unknown>>(
    collection: V,
    iterator: _.CollectionIterator<TypeOfCollection<V>, void, V>,
    context?: unknown): V;
```

but that's probably going too far from a backwards compatibility standpoint. For that matter, 8b9a338 and 5e073e7 are probably too far too, though I really like the idea of Underscore and _Chain being more wrapped-value-centric than collection-item-centric and it does make them a little simpler and more intuitive.

Reply:
<sigh> Too, I feel like these most recent changes will require this to be a single giant PR that has to be reviewed all in one go...so I'll restrain myself for now but I'll always have the commits [over here](https://github.com/reubenrybnik/DefinitelyTyped/tree/underscore-value-centric-interfaces) to help get there later if it becomes desirable to do so. Too, one large PR that's really just a pair of interface changes and a set of trivial return type changes is at least not as bad as a giant PR that's a complete rewrite would be 😄

Reply:
I feel like a breaking change is defensible if it fixes a problem and/or it makes future development much easier. Likewise for huge PRs. The primary reason to (not) do something should be quality, not quantity.

Reply:
I did the bit where I go as far as I'd like to go [here](https://github.com/reubenrybnik/DefinitelyTyped/tree/underscore-value-centric-functions) because it didn't take too long and I'd like to be able to reference it as a potential change later. I think this is more correct from a purely typing perspective, but I'm admittedly not sure I like it from a conceptual perspective. It's probably not great from a performance perspective either. The other potential option to improve type correctness here would be to add a second type parameter V after T, but it also feels a little weird to me to allow T to be defined as anything that extends TypeOfCollection<V> and actually now that I think about it that's probably worse than V in the iterators being a bit off.

Reply:
Fair point @jgonggrijp. Requiring only a collection type T for an interface that wraps a type that happens to involve T and doesn't actually even require the wrapping of a collection at all isn't great in general, something that my futile attempt to make a generic type that is fully backwards compatible with the current generics `<T, V = T>` probably actually emphasizes, so I probably will push for making that change at some point. Too, I have hope that that one won't be a terribly invasive change since there probably aren't tons of instances of folks actually referencing the `Underscore<T, V?>` and `_Chain<T, V?>` types directly.